specs/experimental/cannon-fault-proof-vm-mt.md

```diff
@@ -0,0 +1,511 @@
+For the rest of this doc, we refer to the MTCannon FPVM as simply the FPVM.
+
+Operationally, the FPVM is a state transition function. This state transition is referred to as a _Step_,
+that executes a single instruction. We say the VM is a function $f$, given an input state $S_{pre}$, steps on a
+single instruction encoded in the state to produce a new state $S_{post}$.
+$$f(S_{pre}) \rightarrow S_{post}$$
+
+Thus, the trace of a program executed by the FPVM is an ordered set of VM states.
+
+### New Features
+
+#### Multithreading
+
+MTCannon adds support for [multithreading](https://en.wikipedia.org/wiki/Thread_(computing)).
+Thread management and scheduling are typically handled by the
+[operating system (OS) kernel](https://en.wikipedia.org/wiki/Kernel_%28operating_system%29):
+programs make thread-related requests to the OS kernel via [syscalls](https://en.wikipedia.org/wiki/System_call).
+As such, this implementation includes a few new Linux-specific thread-related [syscalls](#syscalls).
+Additionally, the [FPVM state](#fpvm-state) has been modified in order to track the set of active threads
+and thread-related global state.
+
+#### Robustness
+
+To ensure no unexpected behaviors are triggered, MTCannon will now raise an exception if unrecognized syscalls are
+encountered during program execution.
+
+## Multithreading
+
+The MTCannon FPVM rotates between threads to provide
+[concurrent execution](https://en.wikipedia.org/wiki/Concurrent_computing) rather than
+true [parallel processing](https://en.wikipedia.org/wiki/Parallel_computing).
+The VM state holds an ordered set of thread state objects representing all executing threads.
+
+On any given step, there is one active thread that will be processed.
+
+### Thread Management
+
+The FPVM state contains two thread stacks that are used to represent the set of all threads: `leftThreadStack` and
+`rightThreadStack`. An additional boolean value (`traverseRight`) determines which stack contains the currently active
+thread and how threads are rearranged when the active thread is preempted (see ["Thread Preemption"](#thread-preemption)
+for details).
+
+When traversing right, the thread on the top of the right stack is the active thread, the right stack is referred to as
+the "active" stack,
+and the left the "inactive" stack.  Conversely, when traversing left, the active thread is on top of the left stack,
+the left stack is "active", and the right is "inactive".
+
+Representing the set of threads as two stacks allows for a succinct commitment to the contents of all threads.
+For details, see [“Thread Stack Hashing”](#thread-stack-hashing).
+
+### Thread Traversal Mechanics
+
+Threads are traversed deterministically by moving from the first thread to the last thread,
+then back to the first thread repeatedly.  For example, given the set of threads: {0,1,2,3},
+the FPVM would traverse to each as follows: 0, 1, 2, 3, 2, 1, 0, 1, 2, 3, 2, ….
```

Comment:
each traversal visits the last item in the stack twice. So this should be:
```suggestion
the FPVM would traverse to each as follows: 0, 1, 2, 3, 3, 2, 1, 0, 0, 1, 2, 3, 3, 2, ….
```

Worth clarifying why this happens.

Reply:
good catch!